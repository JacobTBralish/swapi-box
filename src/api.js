const url = 'https://swapi.co/api';

export const fetchApi = async (url) => {
  const fetched = await fetch(url)
  if(fetched.status >= 400) {
    throw(new Error('Error retrieving sick Star Wars Data'))
  }
  else {
     return await fetched.json()
  }
}

export const getFilmNum = () => {
  return Math.floor(Math.random() * 7 + 1)
}

const episodeArray = [
  "Episode IV",
  "Episode V",
  "Episode VI",
  "Episode I",
  "Episode II",
  "Episode III",
  "Episode VII"  
]

export const getScroll = async (num) => {
  const { title, opening_crawl, release_date } = await fetchApi(`${url}/films/${num}`)
  const episodeTitle = episodeArray[ num-1 ]
  return { title, episodeTitle, opening_crawl, release_date }
}

export const getPeople = async () => {
  const { results } = await fetchApi(`${url}/people`)
  const completePeople = getMorePeopleData(results)

  return completePeople
}

const getMorePeopleData = (people) => {
  const completePeople = people.map( async ({ name, homeworld, species }) => {
    homeworld = await fetchApi(homeworld)
    species = await fetchApi(species[0])
    return { 
      Name: name , 
      Homeworld: homeworld.name,
      Population: homeworld.population, 
      Species: species.name
    }
  })

  return Promise.all(completePeople)
}

export const getPlanets = async () => {
  const { results } = await fetchApi(`${url}/planets`);
  const completePlanets = getMorePlanetData(results);

  return completePlanets;
}

const getMorePlanetData = (planets) => {
  const completePlanets = planets.map( async ({ name, residents, terrain, population }) => {
    const retrievedResidents = await getResidents(residents)

    return {
      Name: name, 
      Residents: retrievedResidents, 
      Terrain: terrain, 
      Population: population
    }
  })

  return Promise.all(completePlanets)
}

const getResidents = (residents) => {
  const retrievedResidents = residents.map( async resident => {
    const { name } = await fetchApi(resident)
    
    return name + ' '; 
  })

  return Promise.all(retrievedResidents)
}

export const getVehicles = async () => {
  const { results } = await fetchApi(`${url}/vehicles`)
  const vehicleArray = results.map( ({name, model, vehicle_class, passengers}) => {
    return {
      Name: name, 
      Model: model, 
      Class : vehicle_class, 
      Passengers: passengers
    }
  })

  return vehicleArray;
}
