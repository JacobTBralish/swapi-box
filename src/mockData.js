export const mockPeople = [
    {
      "name": "Luke Skywalker",
      "height": "172",
      "mass": "77",
      "hair_color": "blond",
      "skin_color": "fair",
      "eye_color": "blue",
      "birth_year": "19BBY",
      "gender": "male",
      "homeworld": "https://swapi.co/api/planets/1/",
      "films": [
        "https://swapi.co/api/films/2/",
        "https://swapi.co/api/films/6/",
        "https://swapi.co/api/films/3/",
        "https://swapi.co/api/films/1/",
        "https://swapi.co/api/films/7/"
      ],
      "species": [
        "https://swapi.co/api/species/1/"
      ],
      "vehicles": [
        "https://swapi.co/api/vehicles/14/",
        "https://swapi.co/api/vehicles/30/"
      ],
      "starships": [
        "https://swapi.co/api/starships/12/",
        "https://swapi.co/api/starships/22/"
      ],
      "created": "2014-12-09T13:50:51.644000Z",
      "edited": "2014-12-20T21:17:56.891000Z",
      "url": "https://swapi.co/api/people/1/"
    }
  ]

  export const mockFilm = {
  "title": "A New Hope",
  "episode_id": 4,
  "opening_crawl": "It is a period of civil war.\r\nRebel spaceships, striking\r\nfrom a hidden base, have won\r\ntheir first victory against\r\nthe evil Galactic Empire.\r\n\r\nDuring the battle, Rebel\r\nspies managed to steal secret\r\nplans to the Empire's\r\nultimate weapon, the DEATH\r\nSTAR, an armored space\r\nstation with enough power\r\nto destroy an entire planet.\r\n\r\nPursued by the Empire's\r\nsinister agents, Princess\r\nLeia races home aboard her\r\nstarship, custodian of the\r\nstolen plans that can save her\r\npeople and restore\r\nfreedom to the galaxy....",
  "release_date": "1977-05-25"
}