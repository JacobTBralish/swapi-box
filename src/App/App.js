import React, { Component } from 'react';
import './App.css';
import Header from '../Header/Header';
import Main from '../Main/Main';
import { getScroll } from '../api.js'

class App extends Component {
  constructor() {
    super()

    this.state = {
      scroll: {}
    }
  }

  async componentDidMount() {
    const scroll = await getScroll()
    this.setState({
      scroll
    })
  }

  render() {
    return (
      <div className="App">
        <Header />
        <Main />
      </div>
    );
  }
}

export default App;
