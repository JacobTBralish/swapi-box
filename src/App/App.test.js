import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { mockFilm } from '../mockData'

import * as api from '../api'

describe( 'App Component', () => {
  beforeAll(() => {
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve({
          scroll: mockFilm 
        })
      })
    })
   api.getScroll = () => (mockFilm)
  })

  test('componentDidMount should set the proper state', async () => {
    const expectedState = mockFilm
    const renderedComp = await shallow(<App />);

    await renderedComp.update()

    expect(renderedComp.state('scroll')).toEqual(mockFilm)
  });
});