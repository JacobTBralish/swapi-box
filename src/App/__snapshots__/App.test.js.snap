// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component that it mounts without crashing 1`] = `
<div
  className="App"
>
  <Header />
  <Main
    scroll={Object {}}
  />
</div>
`;
